src/stan/interface_callbacks/writer/stream_writer.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef STAN_INTERFACE_CALLBACKS_WRITER_STREAM_WRITER_HPP
+#define STAN_INTERFACE_CALLBACKS_WRITER_STREAM_WRITER_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <iostream>
+#include <vector>
+#include <string>
+
+namespace stan {
+  namespace interface_callbacks {
+    namespace writer {
+
+      template <class Stream>
```

Comment:
1.  OO: The streams are object oriented and designed to
be used with std::ostream arguments.  The class template
doesn't give you any more flexibility because you need
to instantiate it to a class implementing some concept,
which is going to be defined by ostream itself.

2. References:  If we use std::ostream& arguments,
we don't have to do the == NULL and branch logic
everywhere for safety and we don't have to dereference
the objects in the code.  This assumes anything 
supplying one of these stream-based implementations will 
supply a stream, not NULL (0).

- Bob

> ...

> > +      template <class Stream>
> 
> For many of the same reasons templates are used to replace class derivations. Here it was just to avoid relying on a specific ostream interface (and looking up how sstream and fstream all relate to each other).

>  @betanalpha, any reason we have this templated instead of taking in an std::ostream? — Reply to this email directly or view it on GitHub.